pkg/auth/providers/azuread/azuread_client.go

```diff
@@ -0,0 +1,355 @@
+		return "", httperror.NewAPIError(httperror.MissingRequired, "username or password not provided")
+	}
+	domain := config.Domain
+	endsWith := strings.HasSuffix("username", "@"+domain)
+	if domain != "" && endsWith == false {
+		username = username + "%40" + domain //"@" is encoded to "%40"
+	}
+
+	body := bytes.Buffer{}
+	body.WriteString("scope=openid&grant_type=password&resource=https%3A%2F%2Fgraph.windows.net")
+	body.WriteString("&client_id=")
+	body.WriteString(config.ClientID)
+	body.WriteString("&username=")
+	body.WriteString(username)
+	body.WriteString("&password=")
+	body.WriteString(password)
+
+	url, err := ac.getURL("TOKEN", config, "")
+	if err != nil {
+		logrus.Errorf("AzureAD GET url %v received error from Azure, err: %v", url, err)
+		return "", err
+	}
+
+	resp, err := ac.postToAzureAD(url, body.String())
+	if err != nil {
+		logrus.Errorf("AzureAD getAccessToken: GET url %v received error from Azure, err: %v", url, err)
+		return "", err
+	}
+	defer resp.Body.Close()
+
+	// Decode the response
+	var respMap map[string]interface{}
+	b, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		logrus.Errorf("AzureAD getAccessToken: received error reading response body, err: %v", err)
+		return "", err
+	}
+
+	if err := json.Unmarshal(b, &respMap); err != nil {
+		logrus.Errorf("AzureAD getAccessToken: received error unmarshalling response body, err: %v", err)
+		return "", err
+	}
+
+	if respMap["error"] != nil {
+		desc := respMap["error_description"]
+		logrus.Errorf("Received Error from AzureAD %v, description from AzureAD %v", respMap["error"], desc)
+		return "", fmt.Errorf("Received Error from AzureAD %v, description from AzureAD %v", respMap["error"], desc)
+	}
+
+	acessToken, ok := respMap["access_token"].(string)
+	if !ok {
+		return "", fmt.Errorf("Received Error reading accessToken from response %v", respMap)
+	}
+	return acessToken, nil
+}
+
+func (ac *Client) postToAzureAD(url string, body string) (*http.Response, error) {
+	req, err := http.NewRequest("POST", url, strings.NewReader(body))
+	if err != nil {
+		logrus.Error(err)
+	}
+	req.Header.Add("Accept", "application/json")
+	req.Header.Add("Content-Type", "application/x-www-form-urlencoded")
+
+	resp, err := ac.httpClient.Do(req)
+	if err != nil {
+		logrus.Errorf("Received error from azure: %v", err)
+		return resp, err
+	}
+	// Check the status code
+	switch resp.StatusCode {
+	case 200:
+	case 201:
+	default:
+		var body bytes.Buffer
+		io.Copy(&body, resp.Body)
+		return resp, fmt.Errorf("Request failed, got status code: %d. Response: %s",
+			resp.StatusCode, body.Bytes())
+	}
+	return resp, nil
+}
+
+func (ac *Client) getUser(azureAccessToken string, config *v3.AzureADConfig) (Account, error) {
+	if azureAccessToken == "" {
+		return Account{}, httperror.NewAPIError(httperror.ServerError, "No Azure Access token")
+	}
+	url, err := ac.getURL("USER", config, "")
+
+	resp, err := ac.getFromAzureAD(azureAccessToken, url)
+	if err != nil {
```

Comment:
The 1.6 code checks if response indicates that accessToken has expired (401) and calls refreshAccessToken on it. We'll need to add the same here as right now `getFromAzureAD` only checks 200/201 and returns error for everything else